src/main/java/com/demo/classroom/Security/Filter/JwtAuthenticationFilter.java

```diff
@@ -63,34 +70,57 @@ protected void doFilterInternal(
                     .map(SimpleGrantedAuthority::new)
                     .collect(Collectors.toList());
 
-            Authentication authentication = SecurityContextHolder.getContext().getAuthentication();
+            Authentication authentication = securityContext.getAuthentication();
 
             if (userName != null && (authentication == null || !authentication.getName().equals(userName))) {
                 UserDetails userDetails = this.userDetailsService.loadUserByUsername(userName);
 
                 if (jwtService.isTokenValid(jwt, userDetails)) {
-                    UsernamePasswordAuthenticationToken authToken = new UsernamePasswordAuthenticationToken(
+
+                    var JwtAuthenticationToken= new UsernamePasswordAuthenticationToken(
                             userDetails,
                             null,
                             authorities
                     );
 
-                    authToken.setDetails(new WebAuthenticationDetailsSource().buildDetails(request));
-                    SecurityContextHolder.getContext().setAuthentication(authToken);
-                } else if (refreshToken != null) {
-                    String newAccessToken = jwtService.refreshAccessToken(refreshToken, userDetails);
-                    if (newAccessToken != null) {
-                        response.setHeader("Authorization", "Bearer " + newAccessToken);
-                        response.setHeader("Refresh-Token", refreshToken);
-                    } else {
-                        response.sendError(HttpServletResponse.SC_UNAUTHORIZED, "Refresh token is invalid or expired");
-                        return; 
-                    }
+                    JwtAuthenticationToken.setDetails(new WebAuthenticationDetailsSource().buildDetails(request));
+                    securityContext.setAuthentication(JwtAuthenticationToken);
+                } else {
+                    Optional.ofNullable(refreshToken).ifPresentOrElse(token -> {
+                        String newAccessToken = jwtService.refreshAccessToken(token, userDetails);
+                        if (newAccessToken != null && !newAccessToken.isEmpty()) {
+                            String rotatedRefreshToken = jwtService.rotateRefreshToken(refreshToken, userDetails);
+                            response.setHeader(Constants.AUTH_HEADER, Constants.BEARER_PREFIX + newAccessToken);
+                            response.setHeader(Constants.REFRESH_TOKEN_HEADER, rotatedRefreshToken);
+                        } else {
+                            try {
+                                response.sendError(HttpServletResponse.SC_UNAUTHORIZED, Constants.REFRESH_TOKEN_INVALID.getMessage());
+                            } catch (IOException e) {
+                                e.printStackTrace();
+                            }
```

Comment:
Why are we catching exception here?

Reply:
In here we have `HttpServletResponse response` We can modify the reponse here with corresponding error message and sent it to next filter